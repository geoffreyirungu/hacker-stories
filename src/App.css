.container {
  padding: 20px;
  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);
  color: #171212;
} 
  
.headline-primary {
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
}

.item {
  display: flex;
  align-items: center;
  padding-bottom: 5px;
} 
  
.item > span {
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
} 
.item > span > a {
  color: inherit;
}

.button {
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  transition: all 0.1s ease-in;
}

.button:hover {
  background: #171212;
  color: #ffffff;
} 

.button-small {
  padding: 5px;
} 

.button-large {
  padding: 10px;
}

.button-flex{
  display: flex;
  align-items: center;
}

 .button i{
  padding-left: 2px;
  height:15px;
 }


 .fa-sort-down{
    position: relative;
    top:-12px;
 }
 

.button-more{
  display: block;
  padding: 10px 20px;
  margin: 5px auto;
}

.p-more{
  text-align: center;
}

.search-form {
  padding: 10px 0 20px 0;
  display: flex;
  align-items: baseline;
}

.label {
  border-top: 1px solid #171212;
  border-left: 1px solid #171212;
  padding-left: 5px;
  font-size: 24px;
} 

.input {
  border: none;
  border-bottom: 1px solid #171212;
  background-color: transparent;
  font-size: 24px;
}

.input:focus{
  outline: none;
}

.button:hover > svg > g {
  fill: #ffffff;
  stroke: #ffffff;
}
