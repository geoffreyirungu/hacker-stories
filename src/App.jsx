import * as React from 'react';
import {useState} from 'react';
import axios from 'axios';
import './App.css';
import {SearchForm} from './SearchForm';
import {List} from './List';
import { LastSearches } from './LastSearches';




const useStorageState = (
  key,
  initialState
)=>{
  const[value,setValue]= useState(localStorage.getItem(key) || initialState);

  const isMounted = React.useRef(false);

  React.useEffect(()=>{
    if(!isMounted.current){
      isMounted.current=true;
    }else{
      localStorage.setItem(key,value);
    }
    
  },[key,value]);

  return [value,setValue]
};



const API_BASE = "https://hn.algolia.com/api/v1";
const API_SEARCH="/search";
const PARAM_SEARCH="query=";
const PARAM_PAGE="page=";



const storiesReducer = (state,action)=>{
  switch (action.type){
    case 'STORIES_FETCH_INIT':
      return {...state,isLoading:true,isError:false};
    case 'STORIES_FETCH_FAILURE':
      return {...state,isLoading:false,isError:true};
    case 'STORIES_FETCH_SUCCESS':
      return {
        data: action.payload.page===0 ? action.payload.list: state.data.concat(action.payload.list),
        page: action.payload.page,
        isLoading:false,
        isError:false
      };
    case 'REMOVE_STORY':
      return {...state,data: state.data.filter(story=>story.objectID!==action.payload.objectID)};
    default:
      throw new Error();
  }
};

const getSumComments = (stories) => stories.data.reduce((result,value)=>result + value.num_comments,0);

const extractSearchTerm = url => url.substring(url.lastIndexOf('?')+1,url.lastIndexOf('&'))
  .replace(PARAM_SEARCH,'');

const getLastSearches = (urls) => urls.reduce((result,url,index)=>{
    const searchTerm = extractSearchTerm(url);
    const searchIndex = result.indexOf(searchTerm);
    if(index === 0 || searchIndex == -1)
      return result.concat(searchTerm);

    const lastTerm = result[result.length-1];
    if(searchIndex != -1 && lastTerm != searchTerm)
        return result.filter(s => s != searchTerm).concat(searchTerm);

    return result;

  },[]).slice(-6)
    .slice(0,-1); 

const getUrl = (searchTerm,page) => `${API_BASE}${API_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`;



const App = () =>{

  const[searchTerm,setSearchTerm]=useStorageState("search", 'React');
  const[urls,setUrls]=useState([getUrl(searchTerm,0)]);
  const[stories,dispatchStories]=React.useReducer(storiesReducer,{
    data:[], page:0, isLoading:false, isError:false 
  });

  const lastSearches = getLastSearches(urls);

  const handleSearch = (searchTerm,page)=>{
    const url = getUrl(searchTerm,page);
    setUrls(urls.concat(url));
  };

  const handleLastSearch = (searchTerm) =>{
    setSearchTerm(searchTerm);
    handleSearch(searchTerm,0);
  };

  const handleMore = ()=>{
    const lastUrl = urls[urls.length-1];
    const searchTerm = extractSearchTerm(lastUrl);
    handleSearch(searchTerm,stories.page+1);
  };

  const handleFetchStories = React.useCallback(async ()=>{
    
      dispatchStories({type:'STORIES_FETCH_INIT'});
     try{
        const lastUrl = urls[urls.length-1];
        const result = await axios.get(lastUrl);
        dispatchStories({type:'STORIES_FETCH_SUCCESS',payload: {list: result.data.hits,page: result.data.page}});
     }catch{
        dispatchStories({type:'STORIES_FETCH_FAILURE'});
     }

  },[urls]);

  React.useEffect(()=>{
    handleFetchStories();
  },[handleFetchStories]);


  const handleSearchInput = (event)=>{
    setSearchTerm(event.target.value);
  };


  const handleSearchSubmit = (event) =>{
    handleSearch(searchTerm,0);
    event.preventDefault();
  };

  
  const sumComments = React.useMemo(
    ()=>getSumComments(stories),
  [stories]);

  const handleRemoveStory = React.useCallback(
    (item)=>{
      dispatchStories({type:'REMOVE_STORY',payload:item});
    },[]
  );

  return (
        <div className='container'>
          <h1 className='headline-primary'>Hacker Stories with {sumComments} comments</h1>

          <SearchForm searchTerm={searchTerm} onSearchInput={handleSearchInput} onSearchSubmit={handleSearchSubmit}/>

          <LastSearches lastSearches={lastSearches} onLastSearch={handleLastSearch} />

          {stories.isError && <p>Something went wrong...</p>}

          <List list={stories.data} onRemoveItem={handleRemoveStory}/>

          {
            stories.isLoading ? 
              (<p className='p-more'>Loading...</p>) :
              (<button onClick={handleMore} className='button button-more'>More</button>)
          }

          

        </div>
  );

};
  

export default App;

export {storiesReducer,SearchForm,List};



