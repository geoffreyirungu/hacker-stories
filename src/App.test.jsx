import {describe,it,expect,vi} from 'vitest';
import{render,screen,fireEvent,waitFor} from '@testing-library/react';
import axios from 'axios';

import App, {storiesReducer,Item,List,InputWithLabel,SearchForm} from './App';



vi.mock('axios');

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
    };
    const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};
const stories = [storyOne, storyTwo];
  
describe('stories reducer',()=>{
    it('removes a story from all stories',()=>{
        const state = {data:stories, isLoading:false, isError:false};
        const action = {type:'REMOVE_STORY',payload:storyOne}

        const newState = storiesReducer(state,action);
        const expectedState = {data:[storyTwo], isLoading:false, isError:false};

        expect(newState).toStrictEqual(expectedState);

    })
});


describe('Item',()=>{
    it('renders all properties',()=>{
        render(<Item item={storyOne}/>);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute('href','https://reactjs.org/');
        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('clicking the dismiss button calls callback handler',()=>{
        const handleRemoveItem = vi.fn();
        render(<Item item={storyOne} onRemoveItem={handleRemoveItem}/>);
        fireEvent.click(screen.getByRole('button'));
        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    });


});

describe('SearchForm',()=>{
    const searchFormProps = {
        searchTerm: 'React',
        onSearchSubmit: vi.fn(),
        onSearchInput: vi.fn()
    };

    it('renders the input field with its value',()=>{
        

        render(<SearchForm {...searchFormProps}/>);

        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
    });

    it('renders the correct label',()=>{
        render(<SearchForm {...searchFormProps}/>);

        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
    });

    it('calls onSearchInput when search term changes',()=>{
        render(<SearchForm {...searchFormProps}/>);
        fireEvent.change(screen.getByDisplayValue('React'),{target:{value:'Another value'}});
        expect(searchFormProps.onSearchInput).toBeCalledTimes(1);
    });

    it('calls onSearchSubmit on button submit click',()=>{
        render(<SearchForm {...searchFormProps}/>);
        fireEvent.submit(screen.getByRole('button'));
        expect(searchFormProps.onSearchSubmit).toBeCalledTimes(1);
    });

    it('renders snapshot',()=>{
        const{container}=render(<SearchForm {...searchFormProps}/>);
        expect(container.firstChild).matchSnapshot();
    });

});

describe('App',()=>{

    it('It succeeds fetching data',async ()=>{
        
        const promise = Promise.resolve({
            data:{
                hits: stories
            }
        });

        axios.get.mockImplementationOnce(()=>promise);

        render(<App/>);
        expect(screen.getByText(/Loading/)).toBeInTheDocument();

        await waitFor(async () =>  await promise);
        expect(screen.queryByText(/Loading/)).toBeNull();

        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
        expect(screen.getAllByText('Dismiss').length).toBe(2);

    });

    it('fails fetching data',async ()=>{
        const promise = Promise.reject();

        axios.get.mockImplementationOnce(()=>promise);

        render(<App/>);

        expect(screen.getByText(/Loading/)).toBeInTheDocument();

        try{
            await waitFor(async () =>  await promise);
        }catch{
            expect(screen.queryByText(/Loading/)).toBeNull();
            expect(screen.getByText(/went wrong/)).toBeInTheDocument();
        }

    });

    it('removes a story',async ()=>{
        const promise = Promise.resolve({data:{hits:stories}});
        axios.get.mockImplementationOnce(()=>promise);

        render(<App/>);

        await waitFor(async () => promise);

        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getAllByText('Dismiss').length).toBe(2);

        fireEvent.click(screen.getAllByText('Dismiss')[0]);

        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(screen.getAllByText('Dismiss').length).toBe(1);

    });

    it('searches for specific story',async ()=>{
        const reactPromise = Promise.resolve({data:{hits:stories}});

        const anotherStory = {
            title: 'JavaScript',
            url: 'https://en.wikipedia.org/wiki/JavaScript',
            author: 'Brendan Eich',
            num_comments: 15,
            points: 10,
            objectID: 3,
        };

        const javaScriptPromise = Promise.resolve({data:{hits:[anotherStory]}});

        
        axios.get.mockImplementation((url)=>{
            if(url.includes('React'))
                return reactPromise;
            if(url.includes('JavaScript'))
                return javaScriptPromise;
            throw Error;
        });

        render(<App/>);

        await waitFor(async ()=> reactPromise);

        expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
        expect(screen.queryByDisplayValue('JavaScript')).toBeNull();
        expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.queryByText('Brendan Eich')).toBeNull();

        fireEvent.change(screen.queryByDisplayValue('React'),{target:{value:'JavaScript'}});

        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();

        fireEvent.submit(screen.getByText('Submit'));

        await waitFor(async ()=> javaScriptPromise);

        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();
        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();

    });

    

    

});





