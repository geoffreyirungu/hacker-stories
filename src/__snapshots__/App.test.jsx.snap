// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="search-form"
>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button-large"
    type="submit"
  >
    Submit
  </button>
</form>
`;
